homeassistant/components/efergy/sensor.py

```diff
@@ -94,59 +110,72 @@ async def async_setup_platform(
-        config[CONF_APPTOKEN],
-        async_get_clientsession(hass),
-        utc_offset=config[CONF_UTC_OFFSET],
+    """Set up the Efergy sensor from yaml."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
+        )
     )
 
-    dev = []
-    sensors = await api.get_sids()
-    for variable in config[CONF_MONITORED_VARIABLES]:
-        if variable[CONF_TYPE] == CONF_CURRENT_VALUES:
-            for sensor in sensors:
-                dev.append(
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: entity_platform.AddEntitiesCallback,
+) -> None:
+    """Set up Efergy sensors."""
+    api: Efergy = hass.data[DOMAIN][entry.entry_id][DATA_KEY_API]
+    sids = await api.async_get_sids()
+    sensors = []
+    for description in SENSOR_TYPES:
+        if description.key != CONF_CURRENT_VALUES:
+            sensors.append(
+                EfergySensor(
+                    api,
+                    description,
+                    entry.entry_id,
+                    period=sub("^energy_|^cost_", "", description.key),
+                    currency=hass.config.currency,
+                )
+            )
+        else:
+            description.entity_registry_enabled_default = len(sids) != 1
+            for sid in sids:
+                sensors.append(
                     EfergySensor(
                         api,
-                        variable[CONF_PERIOD],
-                        variable[CONF_CURRENCY],
-                        SENSOR_TYPES[variable[CONF_TYPE]],
-                        sid=sensor["sid"],
+                        description,
+                        entry.entry_id,
+                        sid=sid,
                     )
                 )
-        dev.append(
-            EfergySensor(
-                api,
-                variable[CONF_PERIOD],
-                variable[CONF_CURRENCY],
-                SENSOR_TYPES[variable[CONF_TYPE]],
-            )
-        )
-
-    add_entities(dev, True)
+    async_add_entities(sensors, True)
 
 
-class EfergySensor(SensorEntity):
+class EfergySensor(EfergyEntity, SensorEntity):
     """Implementation of an Efergy sensor."""
 
     def __init__(
         self,
         api: Efergy,
-        period: str,
-        currency: str,
         description: SensorEntityDescription,
+        server_unique_id: str,
+        period: str = None,
+        currency: str = None,
         sid: str = None,
     ) -> None:
         """Initialize the sensor."""
+        super().__init__(api, server_unique_id)
         self.entity_description = description
+        if description.key == CONF_CURRENT_VALUES:
+            self._attr_name = f"{description.name}_{sid}"
+            self._attr_unique_id = f"{server_unique_id}/{description.key}/{sid}"
+        else:
+            self._attr_unique_id = f"{server_unique_id}/{description.key}"
```

Comment:
Why not always include the `sid` in the unique id? The sid will be converted to a string `"None"`in the id if sid is `None`.